sympy/combinatorics/pc_groups.py

```diff
@@ -106,17 +110,18 @@ def minimal_uncollected_subwords(self, word):
             s2, e2 = array[i+1]
             s = ((s1, 1), )
             s = group.dtype(s)
-            if e1 > 0 and e1 < re[s]:
-                if e2 > 0 and index[s1] > index[s2]:
-                    # case-0:  v = x[i]**a*x[i+1], where index[x[i]] > index[x[i+1]]
-                    uncollected_subwords[((s1, e1), (s2, 1))] = 0
-
-                elif e2 < 0 and index[s1] > index[s2]:
-                    # case-1: v = x[i]**a*x[i+1]*-1, where index[x[i]] > index[x[i+1]]
-                    uncollected_subwords[((s1, e1), (s2, -1))] = 1
             if e1 > re[s]-1:
                 # case-2: v = x[i]**a
                 uncollected_subwords[((s1, e1), )] = 2
+                continue
+
+            if e2 > 0 and index[s1] > index[s2]:
+                # case-0:  v = x[i]**a*x[i+1], where index[x[i]] > index[x[i+1]]
+                uncollected_subwords[((s1, e1), (s2, 1))] = 0
+
+            elif e2 < 0 and index[s1] > index[s2]:
```

Comment:
The index condition is the important part. It should be checked first. Then this line and the previous one could be handled together. For example,

    if index[s1] > index[s2]:
        e = 1 if e2 > 0 else -1
        return ((s1, e1), (s2, e))

I don't think it would be necessary to created a dict `uncollected_subwords` when at most one word is returned at a time. Its case can be easily determined afterwards: If the length of the returned tuple is 1, we have a single `((s1, e1),)`. Otherwise we have `((s1, e1), (s2, e))`, and the case can be seen from the sign of `e`.